src/hotspot/share/runtime/serviceThread.cpp

```diff
@@ -59,16 +59,25 @@ JvmtiDeferredEvent* ServiceThread::_jvmti_event = NULL;
 // to add this field to the per-JavaThread event queue.  TODO: fix this sometime later
 JvmtiDeferredEventQueue ServiceThread::_jvmti_service_queue;
 
-// Defer releasing JavaThread OopHandle to the ServiceThread
+// Defer releasing JavaThread OopHandles to the ServiceThread
 class OopHandleList : public CHeapObj<mtInternal> {
-  OopHandle      _handle;
+  static const int _count = 4;  // Need to keep in sync with JavaThread
+  OopHandle _handles[_count];
   OopHandleList* _next;
+  int _index;
  public:
-   OopHandleList(OopHandle h, OopHandleList* next) : _handle(h), _next(next) {}
-   ~OopHandleList() {
-     _handle.release(JavaThread::thread_oop_storage());
-   }
-   OopHandleList* next() const { return _next; }
+  OopHandleList(OopHandleList* next) : _next(next), _index(0) {}
+  void add(OopHandle h) {
+    assert(_index < _count, "too many additions");
+    _handles[_index++] = h;
+  }
+  ~OopHandleList() {
+    assert(_index == _count, "usage error");
+    for (int i = 0; i < _count; i++) {
```

Comment:
Even if they should be the same I think this should be _index.
Since then the code will work fine just adding 3 items (disregarding the asserts).

Reply:
Okay